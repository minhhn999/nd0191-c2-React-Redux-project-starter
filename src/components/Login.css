.imgcontainer {
    text-align: center;
    margin: 24px 0 12px 0;
    position: relative;
}

img.login-image {
    width: 30%;
    height: 30%;
}

.container {
    padding: 16px;
}
input[type=text], input[type=password] {
    width: 100%;
    padding: 12px 20px;
    margin: 8px 0;
    display: inline-block;
    border: 1px solid #ccc;
    box-sizing: border-box;
  }
  
  /* Set a style for all buttons */
  button {
    background-color: #04AA6D;
    color: white;
    padding: 14px 20px;
    margin: 8px 0;
    border: none;
    cursor: pointer;
    width: 100%;
    border-radius: 8px;
  }
  
  button:hover {
    opacity: 0.8;
  }
  