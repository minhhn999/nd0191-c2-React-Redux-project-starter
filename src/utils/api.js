import {
    _getUsers,
    _getQuestions,
    _saveQuestion,
    _login,
  } from './_DATA.js'
  
  export function getInitialData () {
    return Promise.all([
      _getUsers(),
      _getQuestions(),
    ]).then(([users, questions]) => ({
      users,
      questions,
    }))
  }

  export function saveQuestion(question) {
    return _saveQuestion(question)
  }

  export function _saveQuestionAnswer(authedUser, qid, answer) {
    return _saveQuestionAnswer({authedUser, qid, answer})
  }

  export function login(username, password) {
    return Promise.all([
        _getUsers(),
        _getQuestions(),
        _login(username,password)
      ]).then(([users, questions, user]) => ({
        users,
        questions,
        user
      }))
  }

  export function saveQuestionAnswer(authedUser, qid, answer) {
    return _saveQuestionAnswer({authedUser, qid, answer})
  }