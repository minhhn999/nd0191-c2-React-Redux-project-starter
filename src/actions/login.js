import { hideLoading, showLoading } from "react-redux-loading-bar";
import { login } from "../utils/api";
import { receiveUsers } from "./users";
import { receiveQuestions } from "./questions";

export const LOGIN_USER = "LOGIN_USER";
export const LOGOUT_USER = "LOGOUT_USER";

export function loginUser(user) {
  return {
    type: LOGIN_USER,
    user,
  };
}

export function logoutUser() {
  return {
    type: LOGOUT_USER
  };
}

export function handleLogin(username, password) {
  return (dispatch) => {
    dispatch(showLoading());
    return login(username, password)
      .then(({ users, questions, user }) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
        dispatch(loginUser(user));
      })
      .catch((error) => {
        console.error(error);
      })
      .finally(() => {
        dispatch(hideLoading());
      });
  };
}

