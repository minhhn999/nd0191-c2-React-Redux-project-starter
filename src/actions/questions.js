import { showLoading, hideLoading } from "react-redux-loading-bar";
import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { receiveUsers } from "./users";
import { getAuthedUser } from "./login";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTION";
export const ADD_QUESTION = "ADD_QUESTION";
export const ANSWER_QUESTION = "ANSWER_QUESTION";

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}
export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

function answerQuestion(users, questions) {
  return {
    type: ANSWER_QUESTION,
    users,
    questions,
  };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    dispatch(showLoading());
    return saveQuestion({
      optionOneText,
      optionTwoText,
      author: authedUser.id,
    })
      .then((question) => dispatch(addQuestion(question)))
      .catch((error) => {
        console.error(error);
        throw Error(error);
      })
      .finally(() => {
        dispatch(hideLoading());
      });
  };
}
export function handleAnswerQuestion(authedUser, qid, answer) {
  return (dispatch) => {
    dispatch(showLoading());
    return saveQuestionAnswer({ authedUser, qid, answer })
      .then((users, questions) => {
        dispatch(answerQuestion);
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
        dispatch(getAuthedUser(users, authedUser));
      })
      .catch((error) => {
        throw Error(error);
      })
      .finally(() => {
        dispatch(hideLoading());
      });
  };
}
