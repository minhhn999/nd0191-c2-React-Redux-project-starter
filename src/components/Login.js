import React, { useState } from "react";
import employeePollImage from "../assets/employee_poll.png";
import { connect } from "react-redux";
import { handleLogin } from "../actions/login";
import "./Login.css";

const Login = (props) => {
  const [userName, setUserName] = useState("");
  const [password, setPassword] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    console.log(userName, password);
    props.dispatch(handleLogin(userName, password));
  };
  return (
    <form className="modal-content animate" onSubmit={handleSubmit}>
      <div className="imgcontainer">
        <img src={employeePollImage} alt="employee poll" className="login-image" />
      </div>

      <div className="container">
        <label htmlFor="uname">
          <b>Username</b>
        </label>
        <input
          type="text"
          placeholder="Enter Username"
          name="uname"
          required
          onChange={(e) => setUserName(e.target.value)}
        />

        <label htmlFor="psw">
          <b>Password</b>
        </label>
        <input
          type="password"
          placeholder="Enter Password"
          name="psw"
          required
          onChange={(e) => setPassword(e.target.value)}
        />

        <button type="submit">Login</button>
      </div>
    </form>
  );
};

export default connect()(Login);
