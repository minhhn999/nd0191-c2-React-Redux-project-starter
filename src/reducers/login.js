import { LOGIN_USER, LOGOUT_USER } from "../actions/login";

export default function login(state = null, action) {
  switch (action.type) {
    case LOGIN_USER:
      return action.user;
    case LOGOUT_USER:
      return null;
    default:
      return state;
  }
}
