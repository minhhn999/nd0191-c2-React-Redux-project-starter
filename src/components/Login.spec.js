/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable testing-library/prefer-screen-queries */
import React from "react";
import { fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect"; // For additional matchers like toBeInTheDocument

import Login from "./Login";
import { Provider } from "react-redux";
import { handleLogin } from "../actions/login";
import { mockStoreData } from "../utils/mockStoreDate";
import configureStore from "redux-mock-store";
import { act } from "react-dom/test-utils";

jest.mock("../actions/login");
const mockStore = configureStore([]);

describe("Login Component", () => {
  let store;

  beforeEach(() => {
    store = mockStore({ mockStoreData });
    store.dispatch = jest.fn();
  });
  it("will match snapshot", () => {
    var view = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );
    expect(view).toMatchSnapshot();
  });

  it("should update UI after login", async () => {
    const { getByPlaceholderText, getByText } = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );

    act(() => {
      // Fill in the username and password fields
      fireEvent.change(getByPlaceholderText("Enter Username"), {
        target: { value: "testUser" },
      });
      fireEvent.change(getByPlaceholderText("Enter Password"), {
        target: { value: "testPassword" },
      });

      // Click the login button
      fireEvent.click(getByText("Login"));
    });

    // Verify handleLogin is called with correct values
    expect(handleLogin).toHaveBeenCalledWith("testUser", "testPassword");
  });

  it("should render username field, password field, and submit button", () => {
    const { getByPlaceholderText, getByText } = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );

    // Check if username field is present
    const usernameField = getByPlaceholderText("Enter Username");
    expect(usernameField).toBeInTheDocument();

    // Check if password field is present
    const passwordField = getByPlaceholderText("Enter Password");
    expect(passwordField).toBeInTheDocument();

    // Check if submit button is present
    const submitButton = getByText("Login");
    expect(submitButton).toBeInTheDocument();
  });

  it("should display error message if input incorrect username & password", async () => {
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <Provider store={store}>
        <Login />
      </Provider>,
    );

    act(() => {
      // Fill in the username and password fields with incorrect values
      fireEvent.change(getByPlaceholderText("Enter Username"), {
        target: { value: "incorrectUsername" },
      });
      fireEvent.change(getByPlaceholderText("Enter Password"), {
        target: { value: "incorrectPassword" },
      });

      // Click the login button
      fireEvent.click(getByText("Login"));
    });

    await waitFor(() => {
      const errorMessage = getByTestId("error-message");
      expect(errorMessage.textContent).toBe(
        "Incorrect Username or Password. Please retry again!",
      );
    });
  });
});
