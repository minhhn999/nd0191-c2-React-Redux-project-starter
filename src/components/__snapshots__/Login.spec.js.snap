// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="modal-content animate"
      >
        <div
          class="img-container"
        >
          <img
            alt="employee poll"
            class="login-image"
            src="employee_poll.png"
          />
        </div>
        <div
          class="container"
        >
          <label
            for="uname"
          >
            <b>
              Username
            </b>
          </label>
          <input
            name="uname"
            placeholder="Enter Username"
            required=""
            type="text"
          />
          <label
            for="psw"
          >
            <b>
              Password
            </b>
          </label>
          <input
            name="psw"
            placeholder="Enter Password"
            required=""
            type="password"
          />
          <button
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="modal-content animate"
    >
      <div
        class="img-container"
      >
        <img
          alt="employee poll"
          class="login-image"
          src="employee_poll.png"
        />
      </div>
      <div
        class="container"
      >
        <label
          for="uname"
        >
          <b>
            Username
          </b>
        </label>
        <input
          name="uname"
          placeholder="Enter Username"
          required=""
          type="text"
        />
        <label
          for="psw"
        >
          <b>
            Password
          </b>
        </label>
        <input
          name="psw"
          placeholder="Enter Password"
          required=""
          type="password"
        />
        <button
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
